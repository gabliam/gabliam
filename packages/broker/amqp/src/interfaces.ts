import { ConsumeMessage, Options } from 'amqplib';

export interface QueueOptions {
  /**
   * if true, scopes the queue to the connection (defaults to false)
   */
  exclusive: boolean;

  /**
   * if true, the queue will survive broker restarts,
   * modulo the effects of exclusive and autoDelete;
   * this defaults to true if not supplied, unlike the others
   */
  durable: boolean;

  /**
   * if true, the queue will be deleted when the number of consumers drops to zero (defaults to false)
   */
  autoDelete: boolean;

  /**
   * RabbitMQ extensions can also be supplied as options.
   * These typically require non-standard x-* keys and values, sent in the arguments table; e.g., 'x-expires'.
   * When supplied in options, the x- prefix for the key is removed; e.g., 'expires'.
   * Values supplied in options will overwrite any analogous field you put in options.arguments.
   */
  arguments?: { [name: string]: string };
}

export interface QueueConfiguration {
  queueName: string;

  options: QueueOptions;
}

export type ConsumerHandler = (
  msg: ConsumeMessage | null
) => void | Promise<void>;

export type Handler = (...args: any[]) => any | Promise<any>;

export interface Controller {
  [k: string]: Handler;
}

export type ConsumeOptions = Options.Consume;

export type SendOptions = Options.Publish;

export interface RabbitHandlerOptions {
  consumeOptions?: ConsumeOptions;

  sendOptions?: SendOptions;

  sendOptionsError?: SendOptions;
}

export interface ConsumeConfig {
  queueName: string;
  handler: ConsumerHandler;
  options?: ConsumeOptions;
}

export interface QueueDictionnary {
  [k: string]: QueueConfiguration;
}

/**
 * Configuration for a connection
 */
export interface ConnectionConfig {
  /**
   * Name of the connection
   */
  name: string;

  /**
   * Url of the connection
   */
  url: string;

  /**
   * Value passed to rabbitmq for an undefined value
   * amqplib can not send an undefined value because the lib uses Buffer
   *
   */
  undefinedValue: string;

  queues: QueueDictionnary;

  gzipEnabled: boolean;
}
