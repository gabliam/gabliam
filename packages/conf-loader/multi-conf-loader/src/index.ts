import {
  configResolver,
  LoaderConfigParseError,
  Resolver,
} from '@gabliam/core';
import d from 'debug';
import fs from 'fs';
import yaml from 'js-yaml';
import _ from 'lodash';
import { inspect, promisify } from 'util';

// Promisify
const glob = promisify(require('glob')); // require and no import for typings bug
const readFile = promisify(fs.readFile);

const debug = d('@gabliam/multi-conf-loader');

async function loadConstant(
  folder: string,
  projectName: string,
  resolver: Resolver,
  profile?: string,
) {
  const files: string[] = await glob(`**/constants?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let constants: object = {};

  if (!files || files.length === 0) {
    return constants;
  }

  const defaultProfileFile = files.find((file) => file === `constants.yml`);
  if (defaultProfileFile) {
    constants = _.merge(
      {},
      constants,
      await loadFile(`${folder}/${defaultProfileFile}`, resolver),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `constants-${profile}.yml`,
    );

    if (profileFile) {
      constants = _.merge(
        {},
        constants,
        await loadFile(`${folder}/${profileFile}`, resolver),
      );
    }
  }
  return constants;
}

/**
 * FileLoader
 */
const multiConfLoader = async (
  { folder, projectName }: { folder: string; projectName: string },
  profile?: string,
) => {
  debug('loadConfig', folder);
  const resolver = configResolver(folder);
  const constants = await loadConstant(folder, projectName, resolver, profile);

  // create tag config
  const ConfigYamlType = new yaml.Type('!config', {
    kind: 'scalar',
    construct: function (data) {
      return _.get(constants, data, {});
    },
    instanceOf: Object,
  });

  const GAB_SCHEMA = yaml.Schema.create([ConfigYamlType]);

  const files: string[] = await glob(`**/application?(-+([a-zA-Z])).@(yml)`, {
    cwd: folder,
  });

  let config: object = {};

  if (!files || files.length === 0) {
    return config;
  }

  const defaultProfileFile = files.find((file) => file === `application.yml`);
  if (defaultProfileFile) {
    config = _.merge(
      {},
      config,
      await loadConfigProject(
        `${folder}/${defaultProfileFile}`,
        projectName,
        GAB_SCHEMA,
        resolver,
      ),
    );
  }

  if (profile) {
    const profileFile = files.find(
      (file) => file === `application-${profile}.yml`,
    );

    if (profileFile) {
      config = _.merge(
        {},
        config,
        await loadConfigProject(
          `${folder}/${profileFile}`,
          projectName,
          GAB_SCHEMA,
          resolver,
        ),
      );
    }
  }
  debug('loadedConfig', inspect(config, { depth: null }));
  return config;
};

async function loadConfigProject(
  filePath: string,
  projectName: string,
  schema: yaml.Schema,
  resolver: Resolver,
) {
  const data = await readFile(filePath, 'utf8');
  try {
    const vals = yaml.loadAll(data, undefined, { schema }) || [];
    return await resolver(_.find(vals, <any>{ projectName }) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

async function loadFile(filePath: string, resolver: Resolver): Promise<Object> {
  const data = await readFile(filePath, 'utf8');
  try {
    return await resolver(yaml.load(data) || {});
  } catch (e) {
    throw new LoaderConfigParseError(filePath, e);
  }
}

export default multiConfLoader;
