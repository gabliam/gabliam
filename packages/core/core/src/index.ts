import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
