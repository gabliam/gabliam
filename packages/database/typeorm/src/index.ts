import { TypeOrmPlugin } from './typeorm-plugin';

export { ConnectionManager } from './connection-manager';
export * from './connection-options-reader';
export * from './constant';
export * from './typeorm';
export { TypeOrmPlugin as default };
