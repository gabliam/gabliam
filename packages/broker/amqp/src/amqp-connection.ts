import { reflection, toPromise, ValueExtractor } from '@gabliam/core';
import { log4js } from '@gabliam/log4js';
import amqp, {
  AmqpConnectionManager,
  ChannelWrapper,
} from 'amqp-connection-manager';
import { ConfirmChannel, ConsumeMessage, Message } from 'amqplib';
import PromiseB from 'bluebird';
import _ from 'lodash';
import { v4 as uuid } from 'uuid';
import { gunzip, gzip } from 'zlib';
import {
  AmqpConnectionError,
  AmqpMessageIsNullError,
  AmqpQueueDoesntExistError,
  AmqpReplytoIsMissingError,
  AmqpTimeoutError,
} from './errors';
import {
  ConsumeConfig,
  ConsumeOptions,
  ConsumerHandler,
  Controller,
  SendOptions,
} from './interfaces';
import { RabbitHandler, RabbitParamDecorator } from './metadatas';
import { Queue } from './queue';

type ExtractArgsFn = (msg: Message) => Promise<any>;

enum ConnectionState {
  stopped,

  running,

  starting,

  stopping,
}

/**
 * Amqp Connection
 */
export class AmqpConnection {
  private logger = log4js.getLogger(AmqpConnection.name);

  private state = ConnectionState.stopped;

  private connection: AmqpConnectionManager;

  private channel: ChannelWrapper;

  private consumerList: ConsumeConfig[] = [];

  private extractArgs: { [k: string]: ExtractArgsFn } = {};

  constructor(
    public indexConfig: number,
    public name: string,
    private url: string,
    private undefinedValue: string,
    private queues: Queue[],
    private valueExtractor: ValueExtractor,
    private gzipEnabled: boolean,
  ) {}

  /**
   * Start the connection
   */
  async start() {
    if (this.state !== ConnectionState.stopped) {
      return;
    }

    this.state = ConnectionState.starting;
    this.connection = amqp.connect([this.url]);
    this.channel = this.connection.createChannel({
      setup: async (channel: ConfirmChannel) => {
        for (const queue of this.queues) {
          // eslint-disable-next-line no-await-in-loop
          await channel.assertQueue(queue.queueName, queue.queueOptions);
        }
        for (const { queueName, handler, options } of this.consumerList) {
          // eslint-disable-next-line no-await-in-loop
          await channel.consume(queueName, handler, options);
        }
      },
    });

    await new Promise<void>((resolve, reject) => {
      const onConnectFailed = (err: any) => {
        if (
          _.get(err, 'err.errno', undefined) === 'ENOTFOUND' ||
          _.get(err, 'err.code', undefined) === 'ENOTFOUND'
        ) {
          this.channel.removeAllListeners('connect');
          this.channel.removeAllListeners('error');
          reject(err.err);
        } else {
          /* istanbul ignore next */
          this.logger.error(`Amqp error %O`, err);
        }
      };
      this.connection.once('connectFailed', onConnectFailed);

      this.state = ConnectionState.running;
      const isConnect = () => {
        this.connection.removeListener('connectFailed', onConnectFailed);
        resolve();
      };
      this.channel.once('connect', isConnect);
      this.channel.once('error', isConnect);
    });
  }

  /**
   * Add a consumer for a queue
   */
  addConsume(
    queue: string,
    handler: ConsumerHandler,
    options?: ConsumeOptions,
  ) {
    const queueName = this.getQueueName(queue);
    if (!this.queueExist(queueName)) {
      throw new AmqpQueueDoesntExistError(queueName);
    }
    this.consumerList.push({ queueName, handler, options });
  }

  /**
   * contrust consumer with controller instance and HandlerMetadata
   */
  constructAndAddConsume(
    propKey: string,
    handlerMetadata: RabbitHandler,
    controller: Controller,
  ) {
    let consumeHandler: ConsumerHandler;
    if (handlerMetadata.type === 'Listener') {
      consumeHandler = this.constructListener(propKey, controller);
    } else {
      consumeHandler = this.constructConsumer(
        propKey,
        handlerMetadata,
        controller,
      );
    }

    this.addConsume(
      handlerMetadata.queue,
      consumeHandler,
      handlerMetadata.consumeOptions,
    );
  }

  /**
   * Send a content to a queue.
   * Content can be undefined
   */
  async sendToQueue(queue: string, content: any, options?: SendOptions) {
    const queueName = this.getQueueName(queue);
    const channel = this.getChannel();
    await channel.sendToQueue(queueName, await this.contentToBuffer(content), {
      contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
      contentType: 'application/json',
      ...options,
    });
  }

  /**
   * Send a content to a queue and Ack the message
   * Content can be undefined
   */
  async sendToQueueAck(
    queue: string,
    content: any,
    msg: Message,
    options?: SendOptions,
  ) {
    const queueName = this.getQueueName(queue);
    const channel = this.getChannel();
    if (channel === null) {
      /* istanbul ignore next */
      throw new AmqpConnectionError();
    }
    await channel.sendToQueue(queueName, await this.contentToBuffer(content), {
      contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
      contentType: 'application/json',
      ...options,
    });
    await this.channel.ack(msg);
  }

  /**
   * Basic RPC pattern with conversion.
   * Send a Javascrip object converted to a message to a queue and attempt to receive a response, converting that to a Java object.
   * Implementations will normally set the reply-to header to an exclusive queue and wait up for some time limited by a timeout.
   */
  async sendAndReceive<T = any>(
    queue: string,
    content: any,
    options: SendOptions = {},
    timeout = 5000,
  ): Promise<T> {
    let onTimeout = false;
    let chan: ConfirmChannel;
    let replyTo: string;
    let promise = new PromiseB<T>((resolve, reject) => {
      const queueName = this.getQueueName(queue);
      if (!options.correlationId) {
        // eslint-disable-next-line no-param-reassign
        options.correlationId = uuid();
      }
      const correlationId = options.correlationId;

      if (!options.replyTo) {
        // eslint-disable-next-line no-param-reassign
        options.replyTo = `amqpSendAndReceive${uuid()}`;
      }

      if (!options.expiration) {
        // eslint-disable-next-line no-param-reassign
        options.expiration = `${timeout}`;
      }

      replyTo = options.replyTo;
      chan = this.getChannel();

      // create new Queue for get the response
      chan
        .assertQueue(replyTo, {
          exclusive: false,
          autoDelete: true,
          durable: false,
        })
        .then(() =>
          chan.consume(replyTo, async (msg: ConsumeMessage | null) => {
            if (msg === null) {
              /* istanbul ignore next */
              reject(new AmqpMessageIsNullError());
              return;
            }
            if (!onTimeout && msg.properties.correlationId === correlationId) {
              resolve(await this.parseContent(msg));
            }
            chan.ack(msg);

            try {
              await chan.deleteQueue(replyTo);
              // eslint-disable-next-line no-empty
            } catch {}
          }),
        )
        .then(async () => {
          chan.sendToQueue(queueName, await this.contentToBuffer(content), {
            contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
            contentType: 'application/json',
            ...options,
          });
        })
        // catch when error amqp (untestable)
        .catch(
          // prettier-ignore
          /* istanbul ignore next */
          async (err: any) => {
            reject(err);
            if (chan) {
              try {
                await chan.deleteQueue(replyTo);
              // eslint-disable-next-line no-empty
              } catch {}
            }
          },
        );
    });

    if (timeout) {
      promise = promise
        .timeout(timeout)
        .catch(PromiseB.TimeoutError, async (e) => {
          onTimeout = true;
          if (chan) {
            try {
              await chan.deleteQueue(replyTo);
              // eslint-disable-next-line no-empty
            } catch {}
          }
          throw new AmqpTimeoutError((<any>e).message);
        });
    }

    return promise;
  }

  /**
   * Stop the connection
   */
  async stop() {
    if (this.state !== ConnectionState.running) {
      return;
    }
    this.state = ConnectionState.stopping;

    try {
      this.channel.removeAllListeners('connect');
      this.channel.removeAllListeners('error');
      await this.connection.close();
      // eslint-disable-next-line no-empty
    } catch {}

    this.state = ConnectionState.stopped;
  }

  /**
   * Test if queue exist
   */
  queueExist(queueName: string) {
    for (const queue of this.queues) {
      if (queue.queueName === queueName) {
        return true;
      }
    }
    return false;
  }

  /**
   * Get the real queueName
   *
   * Search if the queueName is the index of the map of queues => return queueName
   * Search if the queueName is a key value => return the value
   * else return the queue passed on parameter
   */
  getQueueName(queueName: string) {
    const defaultValue = this.valueExtractor(`"${queueName}"`, queueName);

    if (this.valueExtractor(`application.amqp[0] ? true : false`, false)) {
      return this.valueExtractor(
        `application.amqp[${this.indexConfig}].queues['${queueName}'].queueName`,
        defaultValue,
      );
    }
    return this.valueExtractor(
      `application.amqp.queues['${queueName}'].queueName`,
      defaultValue,
    );
  }

  /**
   * Convert content to buffer for send in queue
   */
  async contentToBuffer(content: any) {
    let data: any;

    if (content === undefined) {
      data = this.undefinedValue;
    } else if (content instanceof Buffer) {
      data = content;
    } else if (typeof content === 'string') {
      data = content;
    } else if (content instanceof Error) {
      data = JSON.stringify(content, Object.getOwnPropertyNames(content));
    } else {
      data = JSON.stringify(content);
    }

    if (this.gzipEnabled) {
      return new Promise<Buffer>((resolve) => {
        gzip(Buffer.from(data), (err, res) => {
          if (err) {
            resolve(Buffer.from(''));
          } else {
            resolve(res);
          }
        });
      });
    }

    return Promise.resolve(Buffer.from(data));
  }

  /**
   * Parse content in message
   */
  async parseContent(msg: Message) {
    let data: any;
    if (this.gzipEnabled) {
      data = await new Promise<any>((resolve) => {
        gunzip(msg.content, (err, res) => {
          if (err) {
            resolve(msg.content.toString());
          } else {
            resolve(res.toString());
          }
        });
      });
    } else {
      data = msg.content.toString();
    }

    try {
      data = JSON.parse(data);
      // eslint-disable-next-line no-empty
    } catch {}

    if (data === this.undefinedValue) {
      return undefined;
    }

    return data;
  }

  private constructListener(
    propKey: string,
    controller: Controller,
  ): ConsumerHandler {
    return async (msg: ConsumeMessage | null) => {
      /* istanbul ignore next */
      if (msg === null) {
        return;
      }

      const extractArgs = this.getExtractArgs(propKey, controller);

      const args = await extractArgs(msg);
      await toPromise(controller[propKey](...args));
      await this.channel.ack(msg);
    };
  }

  private constructConsumer(
    propKey: string,
    handlerMetadata: RabbitHandler,
    controller: Controller,
  ): ConsumerHandler {
    return async (msg: Message | null) => {
      if (msg === null) {
        /* istanbul ignore next */
        return;
      }

      // catch when error amqp (untestable)
      /* istanbul ignore next */
      if (msg.properties.replyTo === undefined) {
        throw new AmqpReplytoIsMissingError();
      }
      const extractArgs = this.getExtractArgs(propKey, controller);
      const args = await extractArgs(msg);

      let response: any;
      let sendOptions: SendOptions;
      try {
        response = await toPromise(controller[propKey](...args));
        sendOptions = handlerMetadata.sendOptions || {};
      } catch (err) {
        response = err;
        sendOptions = handlerMetadata.sendOptionsError || {};
      }

      this.sendToQueueAck(msg.properties.replyTo, response, msg, {
        correlationId: msg.properties.correlationId,
        contentEncoding: this.gzipEnabled ? 'gzip' : undefined,
        contentType: 'application/json',
        ...sendOptions,
      });
    };
  }

  private getExtractArgs(propKey: string, controller: Controller) {
    const k = `${controller.constructor.name}#${propKey}`;
    if (this.extractArgs[k]) {
      return this.extractArgs[k];
    }

    const params = reflection.parameters(<any>controller.constructor, propKey);

    if (params.length === 0) {
      // eslint-disable-next-line no-return-assign
      return (this.extractArgs[k] = async (msg) => [
        await this.parseContent(msg),
      ]);
    }

    const parameters: RabbitParamDecorator[] = params.map(
      (meta) => meta.slice(-1)[0],
    );

    // eslint-disable-next-line no-return-assign
    return (this.extractArgs[k] = async (msg) => {
      const content = await this.parseContent(msg);
      return parameters.map((p) => p.handler(p.args, msg, content));
    });
  }

  private getChannel(): ConfirmChannel {
    // eslint-disable-next-line no-underscore-dangle
    if ((<any>this.channel)._channel === null) {
      throw new AmqpConnectionError();
    }
    // eslint-disable-next-line no-underscore-dangle
    return (<any>this.channel)._channel;
  }
}
