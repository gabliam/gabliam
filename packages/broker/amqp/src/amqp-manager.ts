import { ConnectionConfig, QueueDictionnary } from './interfaces';
import { AmqpConnection } from './amqp-connection';
import { Queue } from './queue';
import { ValueExtractor } from '@gabliam/core';
import {
  AmqpDuplicateConnectionError,
  AmqpConnectionNotFoundError,
} from './errors';

export class AmqpConnectionManager {
  private connections: AmqpConnection[] = [];

  constructor(
    connectionConfigs: ConnectionConfig[],
    valueExtractor: ValueExtractor
  ) {
    for (const [
      index,
      { name, url, queues, undefinedValue, gzipEnabled },
    ] of connectionConfigs.entries()) {
      if (this.connections.find(c => c.name === name) !== undefined) {
        throw new AmqpDuplicateConnectionError(name);
      }

      this.connections.push(
        new AmqpConnection(
          index,
          name,
          url,
          undefinedValue,
          this.createQueue(queues),
          valueExtractor,
          gzipEnabled
        )
      );
    }
  }

  async start() {
    await Promise.all(this.connections.map(c => c.start()));
  }

  async stop() {
    await Promise.all(this.connections.map(c => c.stop()));
  }

  getConnection(name: string) {
    const connection = this.connections.find(c => c.name === name);
    if (!connection) {
      throw new AmqpConnectionNotFoundError(name);
    }
    return connection;
  }

  getDefaultConnection() {
    const connection = this.connections.find(c => c.name === 'default');
    if (connection === undefined) {
      return this.connections[0];
    } else {
      return connection;
    }
  }

  private createQueue(queueConfig: QueueDictionnary = {}) {
    return Object.keys(queueConfig).map<Queue>((k: string) => {
      const q = queueConfig[k];
      return new Queue(q.queueName, q.options);
    });
  }
}
