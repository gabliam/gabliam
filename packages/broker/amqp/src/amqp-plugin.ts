import {
  Container,
  GabliamPlugin,
  Plugin,
  reflection,
  Registry,
  Scan,
} from '@gabliam/core';
import d from 'debug';
import { AmqpConnectionManager } from './amqp-manager';
import { TYPE } from './constants';
import {
  CUnit,
  RabbitConsumer,
  RabbitHandler,
  RabbitListener,
} from './metadatas';

const debug = d('Gabliam:Plugin:AmqpPlugin');

@Plugin()
@Scan()
export class AmqpPlugin implements GabliamPlugin {
  build(container: Container, registry: Registry) {
    debug('build AmqpPlugin');
    const controllerIds = registry.get(TYPE.RabbitController);
    const connectionManager = container.get(AmqpConnectionManager);
    debug('controllerIds', controllerIds, container);
    controllerIds.forEach(({ id: controllerId }) => {
      const controller = container.get<any>(controllerId);
      debug('controller', controller);

      const handlerMetadatas = reflection.propMetadataOfDecorator<RabbitHandler>(
        controller.constructor,
        RabbitConsumer,
        RabbitListener,
      );

      const [cunit] = reflection
        .annotationsOfDecorator<CUnit>(controller.constructor, CUnit)
        .slice(-1);

      if (Object.keys(handlerMetadatas)) {
        for (const [key, metas] of Object.entries(handlerMetadatas)) {
          const [handlerMetadata] = metas.slice(-1);
          if (cunit) {
            connectionManager
              .getConnection(cunit.name)
              .constructAndAddConsume(key, handlerMetadata, controller);
          } else {
            connectionManager
              .getDefaultConnection()
              .constructAndAddConsume(key, handlerMetadata, controller);
          }
        }
      }
    });
  }

  async start(container: Container) {
    const connection = container.get(AmqpConnectionManager);
    await connection.start();
  }

  async stop(container: Container) {
    const connection = container.get(AmqpConnectionManager);
    await connection.stop();
  }
}
