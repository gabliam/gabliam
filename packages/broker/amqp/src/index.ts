import { AmqpPlugin } from './amqp-plugin';

export * from './amqp-connection';
export * from './amqp-manager';
export * from './errors';
export * from './interfaces';
export * from './metadatas';
export * from './queue';
export { AmqpPlugin as default };
