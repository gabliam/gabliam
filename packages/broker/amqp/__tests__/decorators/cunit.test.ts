import { reflection } from '@gabliam/core';
import { CUnit } from '../../src/index';

describe('Cunit decorators', () => {
  test('should add Cunit metadata to a class when decorating a method with @Cunit', () => {
    @CUnit('default')
    class TestEntity {}

    expect(reflection.annotations(TestEntity)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @Cunit', () => {
    expect(function() {
      @CUnit('default')
      @CUnit('default2')
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });
});
