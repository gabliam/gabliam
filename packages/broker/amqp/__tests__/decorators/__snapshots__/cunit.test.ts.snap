// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cunit decorators should add Cunit metadata to a class when decorating a method with @Cunit 1`] = `
[
  DecoratorFactory {
    "name": "default",
  },
]
`;
